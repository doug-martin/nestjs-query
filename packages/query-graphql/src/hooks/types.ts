export enum HookTypes {
  BEFORE_CREATE_ONE = 'BeforeCreateOne',
  BEFORE_CREATE_MANY = 'BeforeCreateMany',
  BEFORE_UPDATE_ONE = 'BeforeUpdateOne',
  BEFORE_UPDATE_MANY = 'BeforeUpdateMany',
  BEFORE_DELETE_ONE = 'BeforeDeleteOne',
  BEFORE_DELETE_MANY = 'BeforeDeleteMany',
  BEFORE_QUERY_MANY = 'BeforeQueryMany',
  BEFORE_FIND_ONE = 'BeforeFindOne',
}
