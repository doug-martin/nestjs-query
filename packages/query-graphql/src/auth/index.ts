export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
