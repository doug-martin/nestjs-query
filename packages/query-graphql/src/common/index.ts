export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
