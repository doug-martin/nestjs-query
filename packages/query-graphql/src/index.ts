export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
