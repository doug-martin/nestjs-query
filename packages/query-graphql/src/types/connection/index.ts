export * from './connection.type';
export { EdgeType, PageInfoType, CursorConnectionType } from './cursor';
export { ArrayConnectionType } from './array-connection.type';
