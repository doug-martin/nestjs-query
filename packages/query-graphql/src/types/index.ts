export {
  PageInfoType,
  EdgeType,
  OffsetPageInfoType,
  ConnectionType,
  ConnectionOptions,
  CursorConnectionType,
  CursorConnectionOptions,
  ArrayConnectionType,
  ArrayConnectionOptions,
  OffsetConnectionType,
  OffsetConnectionOptions,
  StaticConnectionType,
  InferConnectionTypeFromStrategy,
} from './connection';
export { MutationArgsType } from './mutation-args.type';
export {
  FilterType,
  UpdateFilterType,
  DeleteFilterType,
  SubscriptionFilterType,
  QueryArgsType,
  QueryArgsTypeOpts,
  PagingTypes,
  CursorQueryArgsType,
  OffsetQueryArgsType,
  NonePagingQueryArgsType,
  CursorPagingType,
  OffsetPagingType,
  PagingStrategies,
} from './query';
export { CreateManyInputType } from './create-many-input.type';
export { CreateOneInputType } from './create-one-input.type';
export { UpdateManyResponseType } from './update-many-response.type';
export { UpdateOneInputType } from './update-one-input.type';
export { UpdateManyInputType } from './update-many-input.type';
export { DeleteManyResponseType } from './delete-many-reponse.type';
export { DeleteOneInputType } from './delete-one-input.type';
export { DeleteManyInputType } from './delete-many-input.type';
export { ConnectionCursorType, ConnectionCursorScalar } from './cursor.scalar';
export { RelationsInputType } from './relations-input.type';
export { RelationInputType } from './relation-input.type';
export { SubscriptionArgsType } from './subscription-args.type';
export { SubscriptionFilterInputType } from './subscription-filter-input.type';
export { AggregateResponseType, AggregateResponseOpts, AggregateArgsType } from './aggregate';
export { FindOneArgsType } from './find-one-args.type';
