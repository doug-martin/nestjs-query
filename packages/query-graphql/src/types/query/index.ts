export { QueryArgsType } from './query-args.type';
export {
  CursorQueryArgsType,
  OffsetQueryArgsType,
  NonePagingQueryArgsType,
  QueryArgsTypeOpts,
  NonePagingQueryArgsTypeOpts,
  OffsetQueryArgsTypeOpts,
  CursorQueryArgsTypeOpts,
  StaticQueryType,
  QueryType,
} from './query-args';
export {
  FilterType,
  DeleteFilterType,
  UpdateFilterType,
  SubscriptionFilterType,
  AggregateFilterType,
} from './filter.type';
export { CursorPagingType, OffsetPagingType, NonePagingType, PagingTypes, PagingStrategies } from './paging';
