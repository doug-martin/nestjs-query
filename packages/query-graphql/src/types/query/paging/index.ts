export { getOrCreateCursorPagingType } from './cursor-paging.type';
export { getOrCreateOffsetPagingType } from './offset-paging.type';
export { getOrCreateNonePagingType } from './none-paging.type';
export { PagingStrategies } from './constants';
export {
  OffsetPagingType,
  NonePagingType,
  CursorPagingType,
  PagingTypes,
  InferPagingTypeFromStrategy,
} from './interfaces';
