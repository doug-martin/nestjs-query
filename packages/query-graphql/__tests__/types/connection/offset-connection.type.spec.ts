// eslint-disable-next-line max-classes-per-file
import { Field, ObjectType, Query, Resolver } from '@nestjs/graphql';
import { plainToClass } from 'class-transformer';
import { OffsetConnectionType, OffsetPagingType, PagingStrategies } from '../../../src';
import { generateSchema } from '../../__fixtures__';
import { getOrCreateOffsetConnectionType } from '../../../src/types/connection';
import { getOrCreateOffsetPagingType } from '../../../src/types/query/paging';

describe('OffsetConnectionType', (): void => {
  @ObjectType('Test')
  class TestDto {
    @Field()
    stringField!: string;

    @Field()
    numberField!: number;

    @Field()
    boolField!: boolean;
  }

  @ObjectType('TestTotalCount')
  class TestTotalCountDto {
    @Field()
    stringField!: string;
  }

  const createPage = (paging: OffsetPagingType): OffsetPagingType =>
    plainToClass(getOrCreateOffsetPagingType(), paging);

  const createTestDTO = (index: number): TestDto => ({
    stringField: `foo${index}`,
    numberField: index,
    boolField: index % 2 === 0,
  });

  it('should create the connection SDL', async () => {
    const TestConnection = getOrCreateOffsetConnectionType(TestDto, { pagingStrategy: PagingStrategies.OFFSET });
    @Resolver()
    class TestOffsetConnectionTypeResolver {
      @Query(() => TestConnection)
      test(): OffsetConnectionType<TestDto> | undefined {
        return undefined;
      }
    }
    const schema = await generateSchema([TestOffsetConnectionTypeResolver]);
    expect(schema).toMatchSnapshot();
  });

  it('should create the connection SDL with totalCount if enabled', async () => {
    const TestConnectionWithTotalCount = getOrCreateOffsetConnectionType(TestTotalCountDto, {
      pagingStrategy: PagingStrategies.OFFSET,
      enableTotalCount: true,
    });
    @Resolver()
    class TestOffsetConnectionTypeResolver {
      @Query(() => TestConnectionWithTotalCount)
      test(): OffsetConnectionType<TestDto> | undefined {
        return undefined;
      }
    }

    const schema = await generateSchema([TestOffsetConnectionTypeResolver]);
    expect(schema).toMatchSnapshot();
  });

  it('should throw an error if the object is not registered with @nestjs/graphql', () => {
    class TestBadDto {
      @Field()
      stringField!: string;
    }

    expect(() => getOrCreateOffsetConnectionType(TestBadDto, { pagingStrategy: PagingStrategies.OFFSET })).toThrow(
      'Unable to make OffsetConnectionType. Ensure TestBadDto is annotated with @nestjs/graphql @ObjectType',
    );
  });

  describe('limit offset offset connection', () => {
    const TestConnection = getOrCreateOffsetConnectionType(TestDto, { pagingStrategy: PagingStrategies.OFFSET });

    it('should create an empty connection when created with new', () => {
      expect(new TestConnection()).toEqual({
        pageInfo: { hasNextPage: false, hasPreviousPage: false },
        nodes: [],
        totalCountFn: expect.any(Function),
      });
    });

    describe('.createFromPromise', () => {
      it('should create a connections response with an empty query', async () => {
        const queryMany = jest.fn();
        const response = await TestConnection.createFromPromise(queryMany, {});
        expect(queryMany).toHaveBeenCalledTimes(0);
        expect(response).toEqual({
          nodes: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false,
          },
          totalCountFn: expect.any(Function),
        });
      });

      it('should create a connections response with an empty paging', async () => {
        const queryMany = jest.fn();
        const response = await TestConnection.createFromPromise(queryMany, { paging: {} });
        expect(queryMany).toHaveBeenCalledTimes(0);
        expect(response).toEqual({
          nodes: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false,
          },
          totalCountFn: expect.any(Function),
        });
      });

      describe('with limit', () => {
        it('should return hasNextPage and hasPreviousPage false when there are the exact number of records', async () => {
          const queryMany = jest.fn();
          const dtos = [createTestDTO(1), createTestDTO(2)];
          queryMany.mockResolvedValueOnce([...dtos]);
          const response = await TestConnection.createFromPromise(queryMany, { paging: createPage({ limit: 2 }) });
          expect(queryMany).toHaveBeenCalledTimes(1);
          expect(queryMany).toHaveBeenCalledWith({ paging: { limit: 3, offset: 0 } });
          expect(response).toEqual({
            nodes: dtos,
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: false,
            },
            totalCountFn: expect.any(Function),
          });
        });

        it('should return hasNextPage true and hasPreviousPage false when the number of records more than the limit and offset == 0', async () => {
          const queryMany = jest.fn();
          const dtos = [createTestDTO(1), createTestDTO(2), createTestDTO(3)];
          queryMany.mockResolvedValueOnce([...dtos]);
          const response = await TestConnection.createFromPromise(queryMany, { paging: createPage({ limit: 2 }) });
          expect(queryMany).toHaveBeenCalledTimes(1);
          expect(queryMany).toHaveBeenCalledWith({ paging: { limit: 3, offset: 0 } });
          expect(response).toEqual({
            nodes: [dtos[0], dtos[1]],
            pageInfo: {
              hasNextPage: true,
              hasPreviousPage: false,
            },
            totalCountFn: expect.any(Function),
          });
        });
      });

      describe('with offset', () => {
        it('should return hasPreviousPage false if offset == 0', async () => {
          const queryMany = jest.fn();
          const dtos = [createTestDTO(1)];
          queryMany.mockResolvedValueOnce([...dtos]);
          const response = await TestConnection.createFromPromise(queryMany, {
            paging: createPage({ limit: 1, offset: 0 }),
          });
          expect(queryMany).toHaveBeenCalledTimes(1);
          expect(queryMany).toHaveBeenCalledWith({ paging: { limit: 2, offset: 0 } });
          expect(response).toEqual({
            nodes: dtos,
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: false,
            },
            totalCountFn: expect.any(Function),
          });
        });

        it('should return hasPreviousPage true if offset > 0', async () => {
          const queryMany = jest.fn();
          const dtos = [createTestDTO(1), createTestDTO(2), createTestDTO(3)];
          queryMany.mockResolvedValueOnce([...dtos]);
          const response = await TestConnection.createFromPromise(queryMany, {
            paging: createPage({ limit: 2, offset: 1 }),
          });
          expect(queryMany).toHaveBeenCalledTimes(1);
          expect(queryMany).toHaveBeenCalledWith({ paging: { limit: 3, offset: 1 } });
          expect(response).toEqual({
            nodes: [dtos[0], dtos[1]],
            pageInfo: {
              hasNextPage: true,
              hasPreviousPage: true,
            },
            totalCountFn: expect.any(Function),
          });
        });
      });

      it('should create an empty connection', async () => {
        const queryMany = jest.fn();
        queryMany.mockResolvedValueOnce([]);
        const response = await TestConnection.createFromPromise(queryMany, {
          paging: createPage({ limit: 2 }),
        });
        expect(queryMany).toHaveBeenCalledTimes(1);
        expect(queryMany).toHaveBeenCalledWith({ paging: { limit: 3, offset: 0 } });
        expect(response).toEqual({
          nodes: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false,
          },
          totalCountFn: expect.any(Function),
        });
      });
    });
  });
});
