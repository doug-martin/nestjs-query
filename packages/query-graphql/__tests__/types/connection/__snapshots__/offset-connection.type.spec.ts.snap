// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OffsetConnectionType should create the connection SDL 1`] = `
type Test {
  stringField: String!
  numberField: Float!
  boolField: Boolean!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type TestOffsetConnection {
  """Paging information"""
  pageInfo: OffsetPageInfo!

  """Array of nodes."""
  nodes: [Test!]!
}

type Query {
  test: TestOffsetConnection!
}

`;

exports[`OffsetConnectionType should create the connection SDL with totalCount if enabled 1`] = `
type Test {
  stringField: String!
  numberField: Float!
  boolField: Boolean!
}

type TestTotalCount {
  stringField: String!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type TestTotalCountOffsetConnection {
  """Paging information"""
  pageInfo: OffsetPageInfo!

  """Array of nodes."""
  nodes: [TestTotalCount!]!

  """Fetch total count of records"""
  totalCount: Int!
}

type Query {
  test: TestTotalCountOffsetConnection!
}

`;
